### ISSUE TITLE
Keyboard enhancements: Improve tabbing and focus in placeholder blocks

### ISSUE THREAD

**Issue by jasmussen** [WordPress/gutenberg#3016]
In https://github.com/WordPress/gutenberg/issues/2983#issuecomment-335828073 and https://github.com/WordPress/gutenberg/issues/2983#issuecomment-336040541, potential future improvements to tabbing in blocks were discussed. 

Briefly:

- If your text cursor is not in a block, or you don't have any block selected, tabbing through the main content area simply _highlights a block_
- Pressing Enter on a highlighted block _selects_ it, which is the same as setting the text cursor inside a text block, clicking on an image to see the alignment controls, or clicking a button in a placeholder block

When a block is _selected_, you can still tab around the block for the movers and settings shortcuts. See also #3003.

We also have arrowkey navigation (see #2990) across blocks, which behaves/should behave like it would in any textfield. Combined with the slash command, you should be able to build an entire post, only using the keyboard, _without_ having to tab anymore than you absolutely need to, if we make some enhancements. That's what this ticket is about. To explain, here are some flows today, and how we could improve them.

Inserting an image:

![image-now](https://user-images.githubusercontent.com/1204802/31534220-50e0da50-aff6-11e7-962e-0ff929a6da57.gif)

Can we change this so as soon as the image placeholder is inserted, the Upload button has focus, instead of requiring a tab as it does now?

Same with Cover Image:

![cover iamge](https://user-images.githubusercontent.com/1204802/31534266-7ef52a2c-aff6-11e7-80fc-aa3303ed67d6.gif)

Same with Custom HTML:

![custom html](https://user-images.githubusercontent.com/1204802/31534292-9693db9c-aff6-11e7-8e21-4fcb25550078.gif)

Also Classic Text:

![classic](https://user-images.githubusercontent.com/1204802/31534340-cb871936-aff6-11e7-82c4-3919902fe8a2.gif)

Compare this to headings, where pressing tab to set focus isn't necessary:

![heading](https://user-images.githubusercontent.com/1204802/31534374-f2b358e4-aff6-11e7-9c75-422752d1375d.gif)

**Comment by EphoxJames** [comment]
The common element of the ones that focus right away is that they all use Editable.

Editable handles the focus prop by setting the selection to TinyMCE's remembered cursor position normally or at the end if `focus.offset === -1`.

In theory I could do the exact same fix for other blocks - just store the selected components in state and on focus set the selection to them unless offset or some other state would override that.

**Comment by mtias** [comment]
Closing as related to: https://github.com/WordPress/gutenberg/issues/1829
